### PR TITLE
Fix NPE in String find and match methods

### PR DESCRIPTION
**Pull request by pspaude**
Currently null values break the find and match methods leading to unintended results (such as an error instead of MFA prompt in certain situations). Fixes these methods to be resilient.

### CODE REVIEW

**Comment by hdeadman** [comment]
You don't often see catching of NullPointerException, why not check for the data that causes it, log a message and return false?

**Comment by pspaude** [comment]
> You don't often see catching of NullPointerException, why not check for the data that causes it, log a message and return false?

Agree, I would actually like to catch all Exceptions which seems to match the intent, but since I didn't originally create these methods, I'm open to suggestions. If you don't think that's warranted, then yes I can create an if check for null input.

**Comment by hdeadman** [comment]
What was the stack trace (i.e. who was calling the method with null)? Maybe this is a case where we want to fix the calling method to not pass null, and maybe the methods should have their parameters be annotated with `@NonNull` if they aren't going to work with null... Avoiding the NPE or catching exceptions is probably hiding some problem with the calling code....

**Comment by pspaude** [comment]
> What was the stack trace (i.e. who was calling the method with null)? Maybe this is a case where we want to fix the calling method to not pass null, and maybe the methods should have their parameters be annotated with `@NonNull` if they aren't going to work with null... Avoiding the NPE or catching exceptions is probably hiding some problem with the calling code....

org.springframework.webflow.execution.ActionExecutionException: Exception thrown executing org.apereo.cas.web.flow.actions.CheckWebAuthenticationRequestAction@5a787684 in state 'determineDuoRequest' of flow 'mfa-duo' -- action execution attributes were 'map['resolvedAuthenticationEvents' -> list[mfa-duo]]'

                at org.springframework.webflow.execution.ActionExecutor.execute(ActionExecutor.java:62)

                at org.springframework.webflow.action.EvaluateAction.doExecute(EvaluateAction.java:77)

                at org.springframework.webflow.action.AbstractAction.execute(AbstractAction.java:188)

                at org.springframework.webflow.execution.AnnotatedAction.execute(AnnotatedAction.java:145)

                at org.springframework.webflow.execution.ActionExecutor.execute(ActionExecutor.java:51)

                at org.springframework.webflow.engine.ActionState.doEnter(ActionState.java:101)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Transition.execute(Transition.java:228)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.execute(FlowExecutionImpl.java:395)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.execute(RequestControlContextImpl.java:214)

                at org.springframework.webflow.engine.TransitionableState.handleEvent(TransitionableState.java:116)

                at org.springframework.webflow.engine.Flow.handleEvent(Flow.java:547)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.handleEvent(FlowExecutionImpl.java:390)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.handleEvent(RequestControlContextImpl.java:210)

                at org.springframework.webflow.engine.ActionState.doEnter(ActionState.java:105)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Transition.execute(Transition.java:228)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.execute(FlowExecutionImpl.java:395)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.execute(RequestControlContextImpl.java:214)

                at org.springframework.webflow.engine.TransitionableState.handleEvent(TransitionableState.java:116)

                at org.springframework.webflow.engine.Flow.handleEvent(Flow.java:547)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.handleEvent(FlowExecutionImpl.java:390)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.handleEvent(RequestControlContextImpl.java:210)

                at org.springframework.webflow.engine.ActionState.doEnter(ActionState.java:105)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Transition.execute(Transition.java:228)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.execute(FlowExecutionImpl.java:395)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.execute(RequestControlContextImpl.java:214)

                at org.springframework.webflow.engine.TransitionableState.handleEvent(TransitionableState.java:116)

                at org.springframework.webflow.engine.Flow.handleEvent(Flow.java:547)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.handleEvent(FlowExecutionImpl.java:390)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.handleEvent(RequestControlContextImpl.java:210)

                at org.springframework.webflow.engine.ActionState.doEnter(ActionState.java:105)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Transition.execute(Transition.java:228)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.execute(FlowExecutionImpl.java:395)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.execute(RequestControlContextImpl.java:214)

                at org.springframework.webflow.engine.TransitionableState.handleEvent(TransitionableState.java:116)

                at org.springframework.webflow.engine.Flow.handleEvent(Flow.java:547)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.handleEvent(FlowExecutionImpl.java:390)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.handleEvent(RequestControlContextImpl.java:210)

                at org.springframework.webflow.engine.ActionState.doEnter(ActionState.java:105)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Flow.start(Flow.java:527)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.start(FlowExecutionImpl.java:368)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.start(RequestControlContextImpl.java:234)

                at org.springframework.webflow.engine.SubflowState.doEnter(SubflowState.java:101)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Transition.execute(Transition.java:228)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.execute(FlowExecutionImpl.java:395)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.execute(RequestControlContextImpl.java:214)

                at org.springframework.webflow.engine.TransitionableState.handleEvent(TransitionableState.java:116)

                at org.springframework.webflow.engine.Flow.handleEvent(Flow.java:547)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.handleEvent(FlowExecutionImpl.java:390)

                at org.springframework.webflow.engine.impl.RequestControlContextImpl.handleEvent(RequestControlContextImpl.java:210)

                at org.springframework.webflow.engine.ActionState.doEnter(ActionState.java:105)

                at org.springframework.webflow.engine.State.enter(State.java:194)

                at org.springframework.webflow.engine.Flow.start(Flow.java:527)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.start(FlowExecutionImpl.java:368)

                at org.springframework.webflow.engine.impl.FlowExecutionImpl.start(FlowExecutionImpl.java:223)

                at org.springframework.webflow.executor.FlowExecutorImpl.launchExecution(FlowExecutorImpl.java:139)

                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)

                at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)

                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

                at java.base/java.lang.reflect.Method.invoke(Method.java:566)

                at org.springframework.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:282)

                at org.springframework.cloud.context.scope.GenericScope$LockedScopedProxyFactoryBean.invoke(GenericScope.java:499)

                at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)

                at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)

                at com.sun.proxy.$Proxy263.launchExecution(Unknown Source)

                at org.springframework.webflow.mvc.servlet.FlowHandlerAdapter.handle(FlowHandlerAdapter.java:264)

                at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1040)

                at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:943)

                at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)

                at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)

                at javax.servlet.http.HttpServlet.service(HttpServlet.java:626)

                at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)

                at javax.servlet.http.HttpServlet.service(HttpServlet.java:733)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:227)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apereo.cas.web.support.AuthenticationCredentialsThreadLocalBinderClearingFilter.doFilter(AuthenticationCredentialsThreadLocalBinderClearingFilter.java:28)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apereo.cas.web.support.filters.RequestParameterPolicyEnforcementFilter.doFilter(RequestParameterPolicyEnforcementFilter.java:401)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apereo.cas.web.support.filters.ResponseHeadersEnforcementFilter.doFilter(ResponseHeadersEnforcementFilter.java:200)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apereo.cas.web.support.filters.AddResponseHeadersFilter.doFilter(AddResponseHeadersFilter.java:64)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.springframework.security.web.FilterChainProxy.doFilterInternal(FilterChainProxy.java:204)

                at org.springframework.security.web.FilterChainProxy.doFilter(FilterChainProxy.java:183)

                at org.springframework.web.filter.DelegatingFilterProxy.invokeDelegate(DelegatingFilterProxy.java:358)

                at org.springframework.web.filter.DelegatingFilterProxy.doFilter(DelegatingFilterProxy.java:271)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)

                at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)

                at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:93)

                at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apereo.cas.logging.web.ThreadContextMDCServletFilter.doFilter(ThreadContextMDCServletFilter.java:99)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apereo.inspektr.common.web.ClientInfoThreadLocalFilter.doFilter(ClientInfoThreadLocalFilter.java:66)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:126)

                at org.springframework.boot.web.servlet.support.ErrorPageFilter.access$000(ErrorPageFilter.java:64)

                at org.springframework.boot.web.servlet.support.ErrorPageFilter$1.doFilterInternal(ErrorPageFilter.java:101)

                at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)

                at org.springframework.boot.web.servlet.support.ErrorPageFilter.doFilter(ErrorPageFilter.java:119)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)

                at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:119)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apache.logging.log4j.web.Log4jServletFilter.doFilter(Log4jServletFilter.java:71)

                at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:189)

                at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:162)

                at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)

                at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)

                at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)

                at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:143)

                at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)

                at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:687)

                at org.apache.catalina.valves.RemoteIpValve.invoke(RemoteIpValve.java:764)

                at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)

                at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:346)

                at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:374)

                at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)

                at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:887)

                at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1684)

                at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)

                at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)

                at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)

                at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)

                at java.base/java.lang.Thread.run(Thread.java:834)

Caused by: java.lang.NullPointerException

                at java.base/java.util.regex.Matcher.getTextLength(Matcher.java:1770)

                at java.base/java.util.regex.Matcher.reset(Matcher.java:416)

                at java.base/java.util.regex.Matcher.(Matcher.java:253)

                at java.base/java.util.regex.Pattern.matcher(Pattern.java:1133)

                at org.apereo.cas.util.RegexUtils.find(RegexUtils.java:143)

                at org.apereo.cas.web.flow.actions.CheckWebAuthenticationRequestAction.doExecute(CheckWebAuthenticationRequestAction.java:29)

                at org.springframework.webflow.action.AbstractAction.execute(AbstractAction.java:188)

                at jdk.internal.reflect.GeneratedMethodAccessor324.invoke(Unknown Source)

                at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

                at java.base/java.lang.reflect.Method.invoke(Method.java:566)

                at org.springframework.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:282)

                at org.springframework.cloud.context.scope.GenericScope$LockedScopedProxyFactoryBean.invoke(GenericScope.java:499)

                at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)

                at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:212)

                at com.sun.proxy.$Proxy162.execute(Unknown Source)

                at org.springframework.webflow.execution.ActionExecutor.execute(ActionExecutor.java:51)

                ... 150 more

**Comment by hdeadman** [comment]
I think `CheckWebAuthenticationRequestAction` should handle requests with no content type specified. That field isn't mandatory and request.getContentType() returns null if no content type is specified.

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/RegexUtils.java [hunk 1]
```diff
@@ -100,7 +100,12 @@ public static Pattern createPattern(final String pattern, final int flags) {
      * @see Matcher#matches()
      */
     public static boolean matches(final Pattern pattern, final String string) {
-        return pattern.matcher(string).matches();
+        try {
+            return pattern.matcher(string).matches();
+        } catch (NullPointerException npe) {
```
Comment (mmoayyed) [inline core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/RegexUtils.java:105]: Generally never a good idea to catch this. You should be simplify by using something like:

```
pattern != null and string != null and etc
```

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/RegexUtils.java [hunk 2]
```diff
@@ -113,11 +118,16 @@ public static boolean matches(final Pattern pattern, final String string) {
      */
     public static boolean matches(final Pattern pattern, final String value, final boolean completeMatch) {
         val matcher = pattern.matcher(value);
-        LOGGER.debug("Matching value [{}] against pattern [{}]", value, pattern.pattern());
-        if (completeMatch) {
-            return matcher.matches();
+        try {
+            LOGGER.debug("Matching value [{}] against pattern [{}]", value, pattern.pattern());
```
Comment (mmoayyed) [inline core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/RegexUtils.java:122]: Please switch this to trace.

File: core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/RegexUtils.java [hunk 3]
```diff
@@ -129,7 +139,12 @@ public static boolean matches(final Pattern pattern, final String value, final b
      * @see Matcher#find()
      */
     public static boolean find(final Pattern pattern, final String string) {
-        return pattern.matcher(string).find();
+        try {
+            return pattern.matcher(string).find();
```
Comment (mmoayyed) [inline core/cas-server-core-util-api/src/main/java/org/apereo/cas/util/RegexUtils.java:143]: See earlier comment.

File: core/cas-server-core-util/src/test/java/org/apereo/cas/util/RegexUtilsTests.java [hunk 4]
```diff
@@ -36,4 +36,14 @@ public void verifyBlankValidRegex() {
     public void verifyNullRegex() {
         assertFalse(RegexUtils.isValidRegex(null));
     }
+
+    @Test
```
Comment (mmoayyed) [inline core/cas-server-core-util/src/test/java/org/apereo/cas/util/RegexUtilsTests.java:40]: Please also add a test where the pattern itself is null.

**Review by mmoayyed — CHANGES_REQUESTED** [review]
It would be great to also include the stacktrace that caused the NPE so we can investigate the root cause. While the changes here are good defensive measures, they also only treat the symptom. If there is something in the code that is actually passing null and should not be doing it, we ought to review and adjust. (Defensive measures by themselves tend to make troubleshooting more difficult, since you'd never see the NPE and wouldn't know if there is a problem somewhere in the call stack)

**Comment by mmoayyed** [comment]
Thanks for providing the stacktrace. Somehow I missed it when I was doing the review. Sorry about the noise.

**Comment by pbodnar** [comment]
Hi all, just a note: The stacktrace above shows only the case when null `string` parameter is passed to a method - that really should be handled. But it's not too likely that a null `pattern` will be given - I guess such a case doesn't have to be handled here.

Moreover, according to the original comment, it looks like we are talking actually about fixing some _bug_ in MFA at minimum, not just about refactoring. @pspaude, it would be great if you could share the exact scenarios when NPE occur now.

**Comment by hdeadman** [comment]
I think @mmoayyed may have fixed the cause of this error in: https://github.com/apereo/cas/commit/88a62c9ed40d30568cc36f0126f72b089d1d5c9a#diff-b717e0c7da54fc97ea542719d9bb2f9280f0d1109e1d1a75dc6dd1bcec65c45f

There would still be an error thrown if null was passed to `RegexUtils.find` but I don't know if that is a bad thing... I would rather have it error out than hide what was probably a missing check by the caller.
